### DISCUSSION TITLE
Option for Buzzer

### DISCUSSION THREAD

**Discussion by electric-m** [rednblkx/HomeKey-ESP32#39] [Ideas]
Hey there,

I was wondering if it would be possible to add code for using a buzzer at the ESP32.
Something like defining pins, make them configurable in config.h

When a tag is presented, the buzzer should output something like „beep beep“.

michael

**Comment by electric-m** [comment]
Maybe something like this?
Of course the action of pressing g a button would have to be exchanged for the trigger of the NFC chip.

```cpp
void loop() {
    // Read the state of the button value:
    buttonState = digitalRead(buttonPin);
    // Check if the button is pressed.
    // If it is, the buttonState is LOW (because of INPUT_PULLUP) :
    if (buttonState == LOW) E
    // Buzz the butzer 5 times
    for (int i = 0; i < 5; i++) {
      digitalwrite(buzzerPin, HIGH); - //- Turn the Buzzer on delay (300);
      // wait for half a second
      digitalWrite(buzzerPin, LOW);
      // Turn the Buzzer off
      delay(300);
      // Wait for half a second
      ｝
      // Wait for the button to be released
      while (digitalRead(buttonPin) == LOW) {
        delay(30);
        // Little delay for debounce
      }
    }
  ```

**Reply by rednblkx** [reply]
Thanks for the suggestion!
I do see that i would make sense to have some external feedback, I'll see what i can do, i can definitely add an option to the config.h file so it's an opt in for people.

**Reply by 5and0** [reply]
This is my remedial addition which I added... 

`  // Function to activate the buzzer for a short time
void activateBuzzer() {
   // Configure the GPIO pin for output
    printf("Activating buzzer...\n");
    gpio_pad_select_gpio(BUZZER_PIN);
    gpio_set_direction(BUZZER_PIN, GPIO_MODE_OUTPUT);
    // Turn on the buzzer
    gpio_set_level(BUZZER_PIN, 1);
    // Delay for 100 milliseconds
    vTaskDelay(pdMS_TO_TICKS(250));
    // Turn off the buzzer
    gpio_set_level(BUZZER_PIN, 0);
} 
`

I only activate when an NFC tag is scanned as iPhone/Apple Watch provide confirmation....

`      else {
        json payload;
        payload["atqa"] = utils::bufToHexString(atqa, 1, true);
        payload["sak"] = utils::bufToHexString(sak, 1, true);
        payload["uid"] = utils::bufToHexString(uid, uidLen, true);
        payload["homekey"] = false;
        mqtt.publish(MQTT_AUTH_TOPIC, payload.dump().c_str());
        activateBuzzer(); // Activate the buzzer
      }`

I was also thinking it would be good to provide some confirmation of authentication when the tag is accepted/denied. It would be good to have an LED to flash green or red.

**Reply by electric-m** [reply]
I think that’s a good idea, to only activate the buzzer for nfc tags.
Does your buzzer beep only once?

I thought about letting it beep quickly two times when a tag is recognized, because that’s a common sound pattern for „everything is alright, no error“.

**Reply by angelosan** [reply]
Hi,
i am very interested in this solution, please could i have the places where to insert the proposed code.

I added in config.h 

#define BUZZER_PIN 12

i added in main.cpp

// Function to activate the buzzer for a short period of time 
 void activateBuzzer() 
 { 
 // Configures the GPIO pin for the output 
 printf(“Buzzer activation...\n”); 
 gpio_pad_select_gpio(BUZZER_PIN); 
 gpio_set_direction(BUZZER_PIN, GPIO_MODE_OUTPUT);
 // Turn on the buzzer 
 gpio_set_level(BUZZER_PIN, 1);
 // Delay of 100 milliseconds 
 vTaskDelay(pdMS_TO_TICKS(250)); 
 // Turn off the buzzer 
 gpio_set_level(BUZZER_PIN, 0); 
 }

during compilation I get this error

argument of type “int” is incompatible with parameter of type “gpio_num_t ”C/C++(167)
#define BUZZER_PIN 12


I am missing to figure out where to insert

    else { json payload; payload[“atqa”] = utils::bufToHexString(atqa, 1, true); 
        payload[“sak”] = utils::bufToHexString(sak, 1, true); 
        payload[“uid”] = utils::bufToHexString(uid, uidLen, true); payload[“homekey”] = false; 
        mqtt.publish(MQTT_AUTH_TOPIC, payload.dump().c_str()); 
        activateBuzzer(); 
        // Activates the buzzer
       }

Thank you and sorry for my English


Translated with DeepL.com (free version)
